### PR TITLE
Decentralize client & stop using socket.io

### LINKED ISSUE #41: Use WebSockets instead of socket.io, and fix CORS
Blocks #25! Clients currently cannot connect to remote servers not on the same domain because of CORS being disabled.

socket.io seems pretty terrible at sending the correct CORS headers, so it's probably worth transitioning to **ws** on the server and _WebSocket_ in the browser. We don't need the fancy polyfills and fallbacks socket.io implements anyway.

### PR DESCRIPTION
**Pull request by heyitsmeuralex**
Fixes #41 _and_ #25! 🎉

### CODE REVIEW

File: site/index.html [hunk 1]
```diff
@@ -14,10 +14,19 @@
 
 <div id="app">
   <aside id="sidebar">
-    <div class="sidebar-heading">
-      <b>Channels</b>
-      <button class="btn" id="create-channel">New</button>
-    </div>
+    <section class="sidebar-section sidebar-section-light">
+      <div class="sidebar-heading">
+        <div id="server-status">Connecting...</div>
```
Comment (towerofnix) [inline site/index.html]: Should this be "not connected" in the HTML, then set to "Connecting..." by JS?

File: site/js/ChannelsActor.js [hunk 2]
```diff
@@ -118,14 +125,14 @@ export default class ChannelsActor extends Actor {
     return this.channels.find(c => c.name === channelName) || null
   }
 
-  viewChannel(channelID) {
+  async viewChannel(channelID) {
     this.activeChannelID = channelID
-    this.socket.emit('view channel', channelID)
+    //this.socket.emit('view channel', channelID) // TODO use api for this
```
Comment (towerofnix) [inline site/js/ChannelsActor.js]: >use api for this

Keep in mind that channel ID is stored on the server per-socket, not per-user. It actually looks like you've already written the code which handles `view channel`, so I'm wondering why this is commented out.

File: site/js/SessionActor.js [hunk 3]
```diff
@@ -42,22 +61,32 @@ export default class SessionActor extends Actor {
...
   }
 
   isCurrentUser(userID) {
     if (!this.loggedIn) return false
     else return this.sessionObj.user.id === userID
   }
 
+  async switchServer(url) {
+    this.currentServerURL = url
+    this.sessionID = this.sessionIDs[url] || ''
```
Comment (towerofnix) [inline site/js/SessionActor.js]: Good thinking!

File: site/js/Socket.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    if (this.isDead) {
+      return
+    }
+
+    this.isDead = true
+    console.error('socket:: is dead')
+
+    // TODO:
+    // Display some kind of onscreen notification that
+    // the connection to the server has been lost.
```
Comment (towerofnix) [inline site/js/Socket.js]: May want to create an issue for this.

File: site/js/Socket.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        return
+      }
+
+      this.reconnect().then(() => {
+        // Connection re-established! :tada:
+        // (reconnect() automatically sets this.isDead = false)
+        console.info('socket:: connection re-established after death')
+      })
+    }, 2000)
+  }
```
Comment (towerofnix) [inline site/js/Socket.js]: This whole automatic reconnect setup is super handy, thanks.

File: site/js/api.js [hunk 6]
```diff
@@ -1,5 +1,13 @@
...
+// These functions assume that serverURL is running
+// on standard HTTP - this will likely change in the
+// future (e.g. Service Workers only work on https).
+//
+// Note that WebSockets use their own protocols:
+//   http  <--> ws
+//   https <--> wss
+
+export function post(path, dataObj, serverURL) {
+  return fetch('http://' + serverURL + '/api/' + path, {
```
Comment (towerofnix) [inline site/js/api.js]: Could you just fetch `'//' + serverURL`? That way the protocol is automagically picked based on whatever protocol you're using to connect to the client. I guess that wouldn't make `ws://`/`wss://` be automatically picked, though.

File: site/js/index.js [hunk 7]
```diff
@@ -17,20 +18,33 @@ const main = async function() {
...
-  // Actors get references to other actors, plus a
-  // reference to the WebSocket connection.
+  // Actors get references to other actors.
   for (const [ name, actor ] of Object.entries(actors)) {
     actor.name = name
     actor.actors = actors
-    actor.socket = socket
 
-    actor.init() // Actors should subscribe to events here.
+    actor.init() // Actors should subscribe to events from eachother here.
```
Comment (towerofnix) [inline site/js/index.js]: Typo? :package:

File: site/index.html (hunk 1 above)
Comment (heyitsmeuralex) [reply site/index.html]: I guess. The result is the same, however.

File: site/js/ChannelsActor.js (hunk 2 above)
Comment (heyitsmeuralex) [reply site/js/ChannelsActor.js]: Yeah, I uncommented this later afaik...?

Comment (heyitsmeuralex) [reply site/js/ChannelsActor.js]: Oh crap maybe not. It should be uncommented as I reimplemented it!

File: site/js/Socket.js (hunk 5 above)
Comment (heyitsmeuralex) [reply site/js/Socket.js]: It took ages to do, and loads of different ways to do it.

Originally the client pingponged the server (as the server does to the client), but that proved kinda buggy.

File: site/js/api.js (hunk 6 above)
Comment (heyitsmeuralex) [reply site/js/api.js]: We should enforce HTTPS soon anyway - it's needed for #6.

File: site/js/index.js (hunk 7 above)
Comment (heyitsmeuralex) [reply site/js/index.js]: Typo?

Comment (towerofnix) [reply site/js/index.js]: "Eachother" should be two words - "each other".

Comment (heyitsmeuralex) [reply site/js/index.js]: o

Comment (heyitsmeuralex) [reply site/js/index.js]: @towerofnix can you fix this 📦

**Pull request merged**